syslog/syslog.3
.\" $Revision$
.\" Copyright (c) 1985 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms are permitted provided
.\" that: (1) source distributions retain this entire copyright notice and
.\" comment, and (2) distributions including binaries display the following
.\" acknowledgement:  ``This product includes software developed by the
.\" University of California, Berkeley and its contributors'' in the
.\" documentation or other materials provided with the distribution and in
.\" all advertising materials mentioning features or use of this software.
.\" Neither the name of the University nor the names of its contributors may
.\" be used to endorse or promote products derived from this software without
.\" specific prior written permission.
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
.\" WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\"	@(#)syslog.3	6.13 (Berkeley) 6/23/90
.\"
.TH SYSLOG 3 "June 23, 1990"
.UC 5
.SH NAME
syslog, openlog, closelog, setlogmask, vsyslog \- control system log
.SH SYNOPSIS
.B "#include <syslog.h>
.PP
.B "openlog(ident, logopt, facility)
.br
.B "char *ident;
.PP
.B "syslog(priority, message, parameters ... )
.br
.B "int priority;
.br
.B "char *message;
.PP
.B "#include <varargs.h>
.br
.B "vsyslog(priority, message, args)
.br
.B "int priority;
.br
.B "char *message;
.br
.B "va_list args;
.PP
.B "closelog()
.PP
.B "setlogmask(maskpri)
.SH DESCRIPTION
.I Syslog
arranges to write
.I message
onto the system log maintained by
.IR syslogd (8).
The message is tagged with
.IR priority .
The message looks like a
.IR printf (3)
string except that
.B %m
is replaced by the current error message (as referenced by
.IR errno ).
A trailing newline is added if needed.
An alternate form, in which the arguments have already been captured
using the variable-length argument facilities of
.IR varargs (3),
is available under the name
.IR vsyslog .
.PP
This message will be read by
.IR syslogd (8)
and written to the system console, log files, or forwarded to
.I syslogd
on another host as appropriate.
.PP
Priorities are encoded as a
.I facility
and a
.IR level .
The facility describes the part of the system
generating the message.
The level is selected from an ordered list:
.TP
LOG_ALERT
A condition that should be corrected immediately,
such as a corrupted system database.
.TP
LOG_CRIT
Critical conditions,
e.g., hard device errors.
.TP
LOG_DEBUG
Messages that contain information
normally of use only when debugging a program.
.TP
LOG_EMERG
A panic condition.
This is normally broadcast to all users.
.TP
LOG_ERR
Errors.
.TP
LOG_INFO
Informational messages.
.TP
LOG_NOTICE
Conditions that are not error conditions,
but should possibly be handled specially.
.TP
LOG_WARNING
Warning messages.
.PP
If
.I syslog
cannot pass the message to
.IR syslogd ,
it will attempt to write the message to the console (``/dev/console'')
if the LOG_CONS option is set (see below).
.PP
If special processing is needed,
.I openlog
can be called to initialize the log file.
The parameter
.I ident
is a string that is prepended to every message.
.I Logopt
is a bit field indicating logging options.
Current values for
.I logopt
are:
.TP
LOG_CONS
If unable to send the message to
.IR syslogd ,
write it to the console.
.TP
LOG_NDELAY 
Open the connection to
.I syslogd
immediately.
Normally the open is delayed
until the first message is logged.
Useful for programs that need to manage the
order in which file descriptors are allocated.
.TP
LOG_PERROR
Write the message to stderr as well to the system log.
.TP
LOG_PID
Log the process id with each message:
useful for identifying instantiations of daemons.
.PP
The
.I facility
parameter encodes a default facility to be assigned to all messages
that do not have an explicit facility encoded:
.TP
LOG_AUTH
The authorization system:
.IR login (1),
.IR su (1),
.IR getty (8),
etc.
.TP
LOG_CRON
The clock daemon.
.TP
LOG_DAEMON
System daemons, such as
.IR ftpd (8),
.IR routed (8),
etc, that are not provided for explicitly by other facilities.
.TP
LOG_KERN
Messages generated by the kernel.
These cannot be generated by any user processes.
.TP
LOG_LPR
The line printer spooling system:
.IR lpr (1),
.IR lpc (8),
.IR lpd (8),
etc.
.TP
LOG_MAIL
The mail system.
.TP
LOG_NEWS
The network news system.
.TP
LOG_SYSLOG
Messages generated internally by
.IR syslogd (8).
.TP
LOG_USER
Messages generated by random user processes.
This is the default facility identifier if none is specified.
.TP
LOG_UUCP
The uucp system.
.TP
LOG_LOCAL0
Reserved for local use.
Similarly for LOG_LOCAL1 through LOG_LOCAL7.
.PP
.I Closelog
can be used to close the log file.
.PP
.I Setlogmask
sets the log priority mask to
.I maskpri
and returns the previous mask.
Calls to
.I syslog
with a priority not set in
.I maskpri
are rejected.
The mask for an individual priority
.I pri
is calculated by the macro LOG_MASK(\fIpri\fP);
the mask for all priorities up to and including
.I toppri
is given by the macro LOG_UPTO(\fItoppri\fP).
The default allows all priorities to be logged.
.SH EXAMPLES
.nf
syslog(LOG_ALERT, "who: internal error 23");

openlog("ftpd", LOG_PID, LOG_DAEMON);
setlogmask(LOG_UPTO(LOG_ERR));
syslog(LOG_INFO, "Connection from host %d", CallingHost);

syslog(LOG_INFO|LOG_LOCAL2, "foobar error: %m");
.fi
.SH "SEE ALSO"
logger(1),
syslogd(8)


NEWS
Current

    Here are the changes for the latest version of INN.

    * New readers.conf file (replaces nnrp.access) which allows more
      flexible specification of access restrictions.

    * Unified overview has been replaced with an overview API, and there are
      now two separate overview implementations to choose from.  One
      (tradindexed) is very like traditional overview but uses an additional
      index file for faster responses to readers.  The other (buffindexed)
      uses circular buffers and can handle a higher incoming article rate
      while still being fast for readers.  The ovmethod key in inn.conf
      chooses the overview method to use.

    * All article storage and retrieval is now done via the storage API. 
      Traditional spool is now available as a storage type under the storage
      API.

    * The timecaf storage method has been added, similar to timehash but
      storing multiple articles in a single file.  See INSTALL for details
      on it.

    * INN now supports embedded Python filters as well as Perl and TCL
      filters.

Version 2.2

    * New storage.conf file (replaces storage.ctl).

    * New (optional) way of handling non-cancel control messages
      (controlchan) that serializes them and prevents server overload from
      control message storms.

    * Support for actsyncd to fetch active file with ftp; configured by
      default to get the one at ftp://ftp.isc.org/pub/usenet/CONFIG/active.Z
      if you run actsyncd.  Be sure to read the manual page for actsync to
      configure an actsync.ign file for your site, and test simpleftp if you
      do not configure with wget or ncftp.  Also see
      ftp://ftp.isc.org/pub/usenet/CONFIG/README.

    * Some options to configure are now moved to inn.conf (merge-to-groups
      and pgp-verify).

    * inndf, a portable version of df(1) is supplied.

    * New cnfsstat program to show stats of cnfs buffers.

    * news2mail program for gateway'ing news into email is provided.

    * mailpost program for gateway'ing email to news is provided.

    * pullnews program for doing a sucking feed is provided (not meant for
      large feeds).

    * The innshellvars.csh.in script is obsolete (and lives in the obsolete
      directory, for now).



syslog/syslog.c
/*  $Revision$
**  Modified by Rich $alz <rsalz@osf.org> to be more portable to older
**  systems.
*/
/*#define INET_SYSLOG */
/*
 * Copyright (c) 1983, 1988 Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted provided
 * that: (1) source distributions retain this entire copyright notice and
 * comment, and (2) distributions including binaries display the following
 * acknowledgement:  ``This product includes software developed by the
 * University of California, Berkeley and its contributors'' in the
 * documentation or other materials provided with the distribution and in
 * all advertising materials mentioning features or use of this software.
 * Neither the name of the University nor the names of its contributors may
 * be used to endorse or promote products derived from this software without
 * specific prior written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)syslog.c	5.28 (Berkeley) 6/27/90";
#endif /* LIBC_SCCS and not lint */

/*
 * SYSLOG -- print message on log file
 *
 * This routine looks a lot like printf, except that it outputs to the
 * log file instead of the standard output.  Also:
 *	adds a timestamp,
 *	prints the module name in front of the message,
 *	has some other formatting types (or will sometime),
 *	adds a newline on the end of the message.
 *
 * The output of this routine is intended to be read by syslogd(8).
 *
 * Author: Eric Allman
 * Modified to use UNIX domain IPC by Ralph Campbell
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/file.h>
#include "syslog.h"
#if	!defined(INET_SYSLOG)
#include <sys/uio.h>
#else
#include <uio.h>
#include <fcntl.h>
#endif	/* !defined(INET_SYSLOG) */
#include <varargs.h>
#include <stdio.h>
#if	defined(INET_SYSLOG)
#include <netinet/in.h>
#endif	/* defined(INET_SYSLOG) */

#ifndef	_PATH_LOGNAME
    /* if you are running an old syslog, change /dev/log to the same as it
     * is in syslogd.c */
#define	_PATH_LOGNAME	"/dev/log"
#endif
#define index		strchr

#define	_PATH_CONSOLE	"/dev/console"

static int	LogFile = -1;		/* fd for log */
static int	connected;		/* have done connect */
static int	LogStat = 0;		/* status bits, set by openlog() */
static char	*LogTag = "syslog";	/* string to tag the entry with */
static int	LogFacility = LOG_USER;	/* default facility code */

extern char *index(), *strcpy(), *strncpy(), *strcat();
/* =()<extern @<SIZE_T>@	strlen();>()= */
extern size_t	strlen();

/*
**  Return a string representation of errno.
*/
static char *
xstrerror(e)
    int		e;
{
    extern int	sys_nerr;
    extern char	*sys_errlist[];
    static char	buff[30];

    if (e >= 0 && e < sys_nerr)
	return sys_errlist[e];
    (void)sprintf(buff, "Error code %d\n", e);
    return buff;
}

/* VARARGS0 */
syslog(va_alist)
    va_dcl
{
	int pri;
	char *fmt;
	va_list ap;

	va_start(ap);
	pri = va_arg(ap, int);
	fmt = va_arg(ap, char*);
	vsyslog(pri, fmt, ap);
	va_end(ap);
}

vsyslog(pri, fmt, ap)
	int pri;
	register char *fmt;
	va_list ap;
{
	extern int errno;
	register int cnt;
	register char *p;
	time_t now, time();
	int fd, saved_errno;
	char tbuf[2048], fmt_cpy[1024], *stdp, *ctime();

	saved_errno = errno;

	/* see if we should just throw out this message */
	if (!LOG_MASK(LOG_PRI(pri)) || (pri &~ (LOG_PRIMASK|LOG_FACMASK)))
		return;
	if (LogFile < 0 || !connected)
		openlog(LogTag, LogStat | LOG_NDELAY, 0);

	/* set default facility if none specified */
	if ((pri & LOG_FACMASK) == 0)
		pri |= LogFacility;

	/* build the message */
	(void)time(&now);
	(void)sprintf(tbuf, "<%d>%.15s ", pri, ctime(&now) + 4);
	for (p = tbuf; *p; ++p);
	if (LogStat & LOG_PERROR)
		stdp = p;
	if (LogTag) {
		(void)strcpy(p, LogTag);
		for (; *p; ++p);
	}
	if (LogStat & LOG_PID) {
		(void)sprintf(p, "[%ld]", (long) getpid());
		for (; *p; ++p);
	}
	if (LogTag) {
		*p++ = ':';
		*p++ = ' ';
	}

	/* substitute error message for %m */
	{
		register char ch, *t1, *t2;
		char *xstrerror();

		for (t1 = fmt_cpy; ch = *fmt; ++fmt)
			if (ch == '%' && fmt[1] == 'm') {
				++fmt;
				for (t2 = xstrerror(saved_errno);
				    *t1 = *t2++; ++t1);
			}
			else
				*t1++ = ch;
		*t1 = '\0';
	}

	(void)vsprintf(p, fmt_cpy, ap);

	cnt = strlen(tbuf);

	/* output to stderr if requested */
	if (LogStat & LOG_PERROR) {
		struct iovec iov[2];
		register struct iovec *v = iov;

		v->iov_base = stdp;
		v->iov_len = cnt - (stdp - tbuf);
		++v;
		v->iov_base = "\n";
		v->iov_len = 1;
		(void)writev(2, iov, 2);
	}

	/* output the message to the local logger */
	if (send(LogFile, tbuf, cnt, 0) >= 0 || !(LogStat&LOG_CONS))
		return;

	/*
	 * output the message to the console; don't worry about
	 * blocking, if console blocks everything will.
	 */
	if ((fd = open(_PATH_CONSOLE, O_WRONLY, 0)) < 0)
		return;
	(void)strcat(tbuf, "\r\n");
	cnt += 2;
	p = index(tbuf, '>') + 1;
	(void)write(fd, p, cnt - (p - tbuf));
	(void)close(fd);
}

#if	!defined(INET_SYSLOG)
static struct sockaddr SyslogAddr;	/* AF_UNIX address of local logger */
#else
static struct sockaddr_in SyslogAddr;	/* AF_INET address of local logger */
#endif	/* !defined(INET_SYSLOG) */
/*
 * OPENLOG -- open system log
 */
openlog(ident, logstat, logfac)
	char *ident;
	int logstat, logfac;
{
	if (ident != NULL)
		LogTag = ident;
	LogStat = logstat;
	if (logfac != 0 && (logfac &~ LOG_FACMASK) == 0)
		LogFacility = logfac;
	if (LogFile == -1) {
#if	!defined(INET_SYSLOG)
		SyslogAddr.sa_family = AF_UNIX;
		strncpy(SyslogAddr.sa_data, _PATH_LOGNAME,
		    sizeof(SyslogAddr.sa_data));
		if (LogStat & LOG_NDELAY) {
			LogFile = socket(AF_UNIX, SOCK_DGRAM, 0);
			fcntl(LogFile, F_SETFD, 1);
		}
#else
		SyslogAddr.sin_family = AF_INET;
		SyslogAddr.sin_port = htons(514);
		SyslogAddr.sin_addr.s_addr = INADDR_ANY;
  		if (LogStat & LOG_NDELAY) {
			LogFile = socket(AF_INET, SOCK_DGRAM, 0);
  			fcntl(LogFile, F_SETFD, 1);
  		}
#endif	/* !defined(INET_SYSLOG) */
	}
	if (LogFile != -1 && !connected &&
	    connect(LogFile, &SyslogAddr, sizeof(SyslogAddr)) != -1)
		connected = 1;
}

/*
 * CLOSELOG -- close the system log
 */
closelog()
{
	(void) close(LogFile);
	LogFile = -1;
	connected = 0;
}

static int	LogMask = 0xff;		/* mask of priorities to be logged */
/*
 * SETLOGMASK -- set the log mask level
 */
setlogmask(pmask)
	int pmask;
{
	int omask;

	omask = LogMask;
	if (pmask != 0)
		LogMask = pmask;
	return (omask);
}
